File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        if len(nums)==1:
            return nums[0]
        t=[0]*(len(nums))
        t[-1], t[-2] = nums[-1], max(nums[-1], nums[-2])
        
        for i in range(len(nums)-3, -1, -1):
            t[i]=max(nums[i]+t[i+2], t[i+1])
        
        return t[0]
        
        
    
    
################################
# class Solution:
#     def rob(self, nums: List[int]) -> int:
#         def knap(idx):
#             if idx>=len(nums):
#                 return 0
#             if memo[idx]==-1:
#                 memo[idx] = max(knap(idx+2)+nums[idx], knap(idx+1))
#             return memo[idx]
        
#         memo=[-1]*(len(nums)+1)
#         return knap(0)
        
############################
# class Solution:
#     def rob(self, nums: List[int]) -> int:
#         def rec(idx, sumt):
#             if idx>=len(nums):
#                 return sumt
#             return max( rec(idx+2, sumt+nums[idx]), rec(idx+1, sumt) )
        
#         return rec(0, 0)
        
        
        

File: 0198-house-robber/0198-house-robber.java
// class Solution {
//     public int rob(int[] nums) {
//         int last=0,secLast=0;
//         for(int i: nums){
//             int temp=Math.max(i+secLast, last);
//             secLast=last;
//             last=temp;
//         }
//         return last;
//     }
// }




class Solution {
    Integer[] dp;
    public int rob(int[] nums) {
        dp = new Integer[nums.length+1];
        return solve(0, nums);
    }
    int solve(int i, int[] nums){
        if(i>=nums.length) return 0;
        if(dp[i] == null)
            dp[i] = Math.max(solve(i+1, nums), nums[i] + solve(i+2, nums));
        return dp[i];
    }
}




// // wont work
// // testcase [2,1,1,2]
//
// class Solution {
//     public int rob(int[] nums) {
//         int last=0,secLast=0;
//         for(int i: nums){
//             secLast += i;
//             int swap = last;
//             last = secLast;
//             secLast = swap;
//         }
//         return Math.max(last, secLast);
//     }
// }
